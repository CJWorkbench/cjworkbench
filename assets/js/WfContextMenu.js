// Drop-down menu on Workflows List page, for each listed WF
// triggered by click on three-dot icon next to listed workflow

import React from 'react'
import {
  UncontrolledDropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem
} from 'reactstrap'
import PropTypes from 'prop-types'


export default class WfContextMenu extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
       <UncontrolledDropdown>
        <DropdownToggle className='context-button'>
          <div className='context-button--icon icon-more'></div>
        </DropdownToggle>
        <DropdownMenu right className='dropdown-menu'>
          <DropdownItem key={2} onClick={this.props.duplicateWorkflow} className='dropdown-menu-item test-duplicate-button'>
            <i className="icon-duplicate context-menu--item-icon"></i>
            <span className='t-d-gray content-3 ml-3'>Duplicate</span>
          </DropdownItem>
          <DropdownItem key={3} onClick={this.props.deleteWorkflow} className='dropdown-menu-item test-delete-button'>
            <i className="icon-bin context-menu--item-icon"></i>
            <span className='t-d-gray content-3 ml-3'>Delete</span>
          </DropdownItem>
        </DropdownMenu>
       </UncontrolledDropdown>
    );
  }
}

WfContextMenu.propTypes = {
  deleteWorkflow:     PropTypes.func.isRequired,
  duplicateWorkflow:  PropTypes.func.isRequired
};
