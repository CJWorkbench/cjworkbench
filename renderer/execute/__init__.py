from .types import UnneededExecution
from .workflow import execute_workflow


__all__ = ["UnneededExecution", "execute_workflow"]
