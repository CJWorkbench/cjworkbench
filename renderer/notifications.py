from dataclasses import dataclass
import datetime
from typing import Optional
from allauth.account.utils import user_display
from django.contrib.auth.models import User
from django.core.mail import EmailMultiAlternatives
from django.template.loader import render_to_string
from django.conf import settings
from cjwkernel.types import RenderResult
from cjwstate.models import WfModule, Workflow
from server.utils import get_absolute_url
from cjworkbench.i18n.templates import get_i18n_context


@dataclass
class OutputDelta:
    """Description of changes between two versions of WfModule output."""

    user: User
    workflow: Workflow
    wf_module: WfModule
    old_result: Optional[RenderResult]
    new_result: RenderResult

    @property
    def workflow_name(self) -> str:
        return self.workflow.name

    @property
    def wf_module_id(self) -> int:
        return self.wf_module.id

    @property
    def module_name(self) -> str:
        return self.wf_module.module_id_name

    @property
    def workflow_url(self) -> str:
        return get_absolute_url(self.workflow.get_absolute_url())


def email_output_delta(output_delta: OutputDelta, updated_at: datetime.datetime):
    user = output_delta.user

    ctx = {
        **get_i18n_context(user=user),
        "user_name": user_display(user),
        "module_name": output_delta.module_name,
        "workflow_name": output_delta.workflow_name,
        "workflow_url": output_delta.workflow_url,
        "date": updated_at.strftime("%b %-d, %Y at %-I:%M %p"),
    }
    subject = render_to_string("notifications/new_data_version_subject.txt", ctx)
    subject = "".join(subject.splitlines())
    message = render_to_string("notifications/new_data_version.txt", ctx)
    mail = EmailMultiAlternatives(
        subject=subject,
        body=message,
        from_email=settings.DEFAULT_FROM_EMAIL,
        to=[user.email],
    )
    mail.attach_alternative(message, "text/html")
    mail.send()
