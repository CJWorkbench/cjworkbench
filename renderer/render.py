import asyncio
from enum import Enum
import logging
import os
from typing import Any, Awaitable, Callable, Dict
from django.db import DatabaseError, InterfaceError
from cjworkbench import rabbitmq
from cjworkbench.pg_render_locker import PgRenderLocker, WorkflowAlreadyLocked
from cjworkbench.sync import database_sync_to_async
from cjworkbench.util import benchmark
from cjwstate.models import Workflow
from . import execute


logger = logging.getLogger(__name__)


@database_sync_to_async
def _lookup_workflow(workflow_id: int) -> Workflow:
    """Lookup workflow, or raise Workflow.DoesNotExist."""
    return Workflow.objects.get(id=workflow_id)


class RenderResult(Enum):
    CHECK_TO_REQUEUE = 1
    MUST_REQUEUE = 2
    MUST_NOT_REQUEUE = 3


async def render_workflow_once(workflow: Workflow, delta_id: int):
    """
    Render a workflow, returning `RenderResult`.
    
    Considers all conceivable errors:

    * Treat UnneededExecution as success -- it's like a render, only faster!
      Always requeue, regardless of the workflow's delta: if the user does
      something and hits undo quickly, the delta ID is the same and yet we want
      to render again.
    * Raise asyncio.CancelledError if there was a cancellation. (This is always
      a bug -- we can't handle cancellation, and it isn't worth our effort.)
    * Re-raise DatabaseError, InterfaceError -- our caller should die when it
      sees these.
    * Catch _every other exception_ (!!!); email us. Return MUST_NOT_REQUEUE:
      we want to leave the workflow in an indeterminate state rather than cause
      this error over and over again.
    """
    # We don't use `workflow.cooperative_lock()` because `execute` may
    # take ages (and it locks internally when it needs to).
    # `execute_workflow()` _anticipates_ that `workflow` data may be
    # stale.
    try:
        task = execute.execute_workflow(workflow, delta_id)
        await benchmark(logger, task, "execute_workflow(%d, %d)", workflow.id, delta_id)
        return RenderResult.CHECK_TO_REQUEUE
    except execute.UnneededExecution:
        logger.info(
            "UnneededExecution in execute_workflow(%d, %d)", workflow.id, delta_id
        )
        return RenderResult.MUST_REQUEUE
    except asyncio.CancelledError:
        raise
    except (DatabaseError, InterfaceError):
        # handled in outer try (which also handles PgRenderLocker)
        raise
    except Exception:
        logger.exception("Error during render of workflow %d", workflow.id)
        return RenderResult.MUST_NOT_REQUEUE


async def render_workflow_and_maybe_requeue(
    pg_render_locker: PgRenderLocker,
    workflow_id: int,
    delta_id: int,
    ack: Callable[[], Awaitable[None]],
    requeue: Callable[[int, int], Awaitable[None]],
) -> None:
    """
    Acquire an advisory lock and render, or re-queue task if the lock is held.

    If a render is requested on a Workflow that's already being rendered,
    there's no point in wasting CPU cycles starting from scratch. Wait for the
    first render to exit (which will happen at the next stale database-write).
    It should then re-schedule a render.
    """
    # Query for workflow before locking. We don't need a lock for this, and no
    # lock means we can dismiss spurious renders sooner, so they don't fill the
    # render queue.
    try:
        workflow = await _lookup_workflow(workflow_id)
    except Workflow.DoesNotExist:
        logger.info("Skipping render of deleted Workflow %d", workflow_id)
        await ack()
        return

    try:
        async with pg_render_locker.render_lock(workflow_id) as lock:
            try:
                result = await render_workflow_once(workflow, delta_id)
            except (asyncio.CancelledError, DatabaseError, InterfaceError):
                raise  # all undefined behavior

            # requeue if needed
            await lock.stall_others()
            if result == RenderResult.MUST_REQUEUE:
                want_requeue = True
            elif result == RenderResult.MUST_NOT_REQUEUE:
                want_requeue = False
            else:
                try:
                    workflow = await _lookup_workflow(workflow_id)
                    if workflow.last_delta_id != delta_id:
                        logger.info(
                            "Requeueing render(workflow=%d, delta=%d)",
                            workflow_id,
                            workflow.last_delta_id,
                        )
                        want_requeue = True
                    else:
                        want_requeue = False
                except Workflow.DoesNotExist:
                    logger.info("Skipping requeue of deleted Workflow %d", workflow_id)
                    want_requeue = False
            if want_requeue:
                await requeue(workflow_id, workflow.last_delta_id)
                # This is why we used `lock.stall_others()`: after requeue,
                # another renderer may try to lock this workflow and we want
                # that lock to _succeed_ -- not raise WorkflowAlreadyLocked.
            # Only ack() _after_ requeue. That preserves our invariant: if we
            # schedule a render, there is always an un-acked render for that
            # workflow queued in RabbitMQ until the workflow is up-to-date. (At
            # this exact moment, there are briefly two un-acked renders.)
            await ack()
    except WorkflowAlreadyLocked:
        logger.info("Workflow %d is being rendered elsewhere; ignoring", workflow_id)
        await ack()
    except (DatabaseError, InterfaceError):
        # Possibilities:
        #
        # 1. There's a bug in renderer.execute. This may leave the event
        # loop's executor thread's database connection in an inconsistent
        # state. [2018-11-06 saw this on production.] The best way to clear
        # up the leaked, broken connection is to die. (Our parent process
        # should restart us, and RabbitMQ will give the job to someone
        # else.)
        #
        # 2. The database connection died (e.g., Postgres went away). The
        # best way to clear up the leaked, broken connection is to die.
        # (Our parent process should restart us, and RabbitMQ will give the
        # job to someone else.)
        #
        # 3. PgRenderLocker's database connection died (e.g., Postgres went
        # away). We haven't seen this much in practice; so let's die and let
        # the parent process restart us.
        #
        # 4. There's some design flaw we haven't thought of, and we
        # shouldn't ever render this workflow. If this is the case, we're
        # doomed.
        #
        # If you're seeing this error that means there's a bug somewhere
        # _else_. If you're staring at a case-3 situation, please remember
        # that cases 1 and 2 are important, too.
        logger.exception("Fatal database error; exiting")
        os._exit(1)


async def _queue_render(workflow_id, delta_id):
    # a separate function for dependency injection
    connection = rabbitmq.get_connection()
    await connection.queue_render(workflow_id, delta_id)


async def handle_render(
    message: Dict[str, Any],
    ack: Callable[[], Awaitable[None]],
    pg_render_locker: PgRenderLocker,
) -> None:
    try:
        workflow_id = int(message["workflow_id"])
        delta_id = int(message["delta_id"])
    except asyncio.CancelledError:
        raise
    except Exception:
        logger.info(
            (
                "Ignoring invalid render request. "
                "Expected {workflow_id:int, delta_id:int}; got %r"
            ),
            message,
        )
        return

    await render_workflow_and_maybe_requeue(
        pg_render_locker, workflow_id, delta_id, ack, _queue_render
    )
