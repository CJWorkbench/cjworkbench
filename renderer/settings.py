import os

from cjworkbench.settings.database import *
from cjworkbench.settings.logging import *
from cjworkbench.settings.rabbitmq import *
from cjworkbench.settings.smtp import *
from cjworkbench.settings.s3 import *

SITE_ID = 1  # for finding domain name when sending emails

# Renderer uses asyncio because it uses RabbitMQ. But when it comes to the
# database, for all intents and purposes it's single-threaded.
N_SYNC_DATABASE_CONNECTIONS = 1

INSTALLED_APPS = [
    "django.contrib.auth",  # cjwstate.models.workflow imports User
    "django.contrib.contenttypes",
    "django.contrib.sites",
    "cjworkbench",  # Workflow model, etc.
]

MIDDLEWARE = []

SECRET_KEY = "not-a-web-server"

# TEMPLATES for email notifications
TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": ["/app/renderer/templates"],
    }
]

API_URL = os.environ["CJW_API_URL"]
"""Where the API server is, from users' perspective.

For example: https://api.workbenchdata.com

The renderer does not interact with the API server; but it writes its URL into
datapackages.
"""
