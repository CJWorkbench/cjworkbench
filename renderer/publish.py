import datetime
import gzip
import hashlib
import json
import shutil
import subprocess
from pathlib import Path
from typing import Any, Dict, List

import cjwparquet
import pyarrow as pa
from django.conf import settings
from django.utils.text import slugify

from cjwkernel.util import json_encode, tempfile_context
from cjwkernel.types import Column, ColumnType
from cjwstate import s3
from .execute.types import TabResult


DELAY_FROM_DATASET_EXPIRED_TO_DELETED = datetime.timedelta(days=1)


def _get_latest_revision(workflow_id: int) -> int:
    """Return the latest revision, or raise FileNotFoundError."""
    with s3.temporarily_download(
        s3.DatasetsBucket, f"wf-{workflow_id}/datapackage.json"
    ) as path, open(path, "rb") as f:
        return json.load(f)["_workbenchRevision"]


def _arrow_type_to_frictionless_field_type(arrow_type: pa.DataType) -> str:
    if pa.types.is_timestamp(arrow_type):
        return "datetime"
    elif pa.types.is_date32(arrow_type):
        return "date"
    elif pa.types.is_integer(arrow_type) or pa.types.is_floating(arrow_type):
        return "number"
    elif (
        pa.types.is_dictionary(arrow_type)
        and pa.types.is_unicode(arrow_type.value_type)
    ) or pa.types.is_unicode(arrow_type):
        return "string"
    else:
        raise ValueError("Unexpected arrow_type %r" % (arrow_type,))


def _build_frictionless_table_schema(schema: pa.Schema) -> Dict[str, Any]:
    return dict(
        fields=[
            dict(name=name, type=_arrow_type_to_frictionless_field_type(arrow_type))
            for name, arrow_type in zip(schema.names, schema.types)
        ]
    )


def _md5sum(path: Path) -> str:
    BLOCK_SIZE = 10 * 1024 * 1024  # arbitrary
    m = hashlib.md5()
    with open(path, "rb") as f:
        while (block := f.read(BLOCK_SIZE)) != b"":
            m.update(block)
    return m.hexdigest()


def _publish_tab_parquet_resource(
    *,
    url_prefix: str,
    s3_prefix: str,
    slug: str,
    title: str,
    parquet_path: Path,
    arrow_schema: pa.Schema,
):
    inner_path = f"/data/{slug}_parquet.parquet"
    s3.fput_file(s3.DatasetsBucket, s3_prefix + inner_path, parquet_path)
    md5sum = _md5sum(parquet_path)
    size = parquet_path.stat().st_size

    return dict(
        profile="data-resource",
        name=slug + "_parquet",
        path=url_prefix + inner_path,
        title=title,
        format="parquet",
        schema=_build_frictionless_table_schema(arrow_schema),
        hash=md5sum,
        bytes=size,
    )


def run_subprocess_and_pipe_stdout_to_file_object(cmd, *, stdout, **kwargs):
    """Like subprocess.run(cmd, stdout=stdout, **kwargs) ... with a file object.

    subprocess.run() will normally pass a file descriptor to the subprocess.
    We need special code to pass a file *object*: we need to read in a loop.
    """
    if "stdin" in kwargs or "stderr" in kwargs:
        # This would be too complicated.
        raise TypeError("Cannot stream from subprocess if stdin or stderr are provided")

    with subprocess.Popen(cmd, stdout=subprocess.PIPE, bufsize=0, **kwargs) as proc:
        shutil.copyfileobj(proc.stdout, stdout)
        proc.wait()

        if proc.returncode != 0:
            raise RuntimeError(
                "%r returned non-zero returncode: %d" % (args, proc.returncode)
            )


def _publish_tab_csv_resource(
    *,
    url_prefix: str,
    s3_prefix: str,
    slug: str,
    title: str,
    arrow_schema: pa.Schema,
    parquet_path: Path,
):
    inner_path = f"/data/{slug}_csv.csv.gz"

    with tempfile_context(suffix=".csv.gz") as csv_path:
        with open(csv_path, "wb", buffering=0) as wf, gzip.GzipFile(
            filename=f"{slug}_csv.csv",
            mode="wb",
            fileobj=wf,
            compresslevel=3,
            mtime=0,
        ) as wzf:
            run_subprocess_and_pipe_stdout_to_file_object(
                ["/usr/bin/parquet-to-text-stream", parquet_path, "csv"],
                stdout=wzf,
            )

        s3.fput_file(s3.DatasetsBucket, s3_prefix + inner_path, csv_path)
        md5sum = _md5sum(csv_path)
        size = csv_path.stat().st_size
    return dict(
        profile="tabular-data-resource",
        name=slug + "_csv",
        path=url_prefix + inner_path,
        title=title,
        format="csv",
        compression="gz",
        schema=_build_frictionless_table_schema(arrow_schema),
        hash=md5sum,
        bytes=size,
    )


def _publish_tab_json_resource(
    *,
    url_prefix: str,
    s3_prefix: str,
    slug: str,
    title: str,
    arrow_schema: pa.Schema,
    parquet_path: Path,
):
    inner_path = f"/data/{slug}_json.json.gz"

    with tempfile_context(suffix=".json") as json_path:
        with json_path.open("wb") as wf, gzip.GzipFile(
            filename=f"{slug}_json.json",
            mode="wb",
            fileobj=wf,
            compresslevel=3,
            mtime=0,
        ) as wzf:
            run_subprocess_and_pipe_stdout_to_file_object(
                ["/usr/bin/parquet-to-text-stream", parquet_path, "json"],
                stdout=wzf,
            )

        s3.fput_file(s3.DatasetsBucket, s3_prefix + inner_path, json_path)
        md5sum = _md5sum(json_path)
        size = json_path.stat().st_size
    return dict(
        profile="data-resource",
        name=slug + "_json",
        path=url_prefix + inner_path,
        title=title,
        format="json",
        compression="gz",
        schema=_build_frictionless_table_schema(arrow_schema),
        hash=md5sum,
        bytes=size,
    )


def _publish_tab_resources(
    *, url_prefix: str, s3_prefix: str, tab_result: TabResult
) -> List[Dict[str, Any]]:
    slug = slugify(tab_result.tab_name)
    with tempfile_context(suffix=".parquet") as parquet_path:
        with pa.ipc.open_file(tab_result.path) as arrow_reader:
            arrow_table = arrow_reader.read_all()
            arrow_schema = arrow_table.schema

            cjwparquet.write(parquet_path, arrow_table)

        parquet_resource = _publish_tab_parquet_resource(
            url_prefix=url_prefix,
            s3_prefix=s3_prefix,
            slug=slug,
            title=tab_result.tab_name,
            arrow_schema=arrow_schema,
            parquet_path=parquet_path,
        )
        csv_resource = _publish_tab_csv_resource(
            url_prefix=url_prefix,
            s3_prefix=s3_prefix,
            slug=slug,
            title=tab_result.tab_name,
            arrow_schema=arrow_schema,
            parquet_path=parquet_path,
        )
        json_resource = _publish_tab_json_resource(
            url_prefix=url_prefix,
            s3_prefix=s3_prefix,
            slug=slug,
            title=tab_result.tab_name,
            arrow_schema=arrow_schema,
            parquet_path=parquet_path,
        )
    return [parquet_resource, csv_resource, json_resource]


def _publish_datapackage(
    *,
    url_prefix: str,
    s3_prefix: str,
    workflow_id: int,
    workflow_name: str,
    revision: int,
    resources: List[Dict[str, Any]],
) -> Dict[str, Any]:
    package_name = f"{workflow_id}-{slugify(workflow_name)}"
    datapackage = dict(
        profile="data-package",
        version=f"0.1.{revision}",
        path=url_prefix + "/datapackage.json",
        title=workflow_name,
        name=package_name,
        created=datetime.datetime.utcnow().isoformat() + "Z",
        resources=resources,
        _workbenchRevision=revision,
    )
    contents = json_encode(datapackage).encode("utf-8")
    s3.put_bytes(s3.DatasetsBucket, s3_prefix + "/datapackage.json", contents)
    return datapackage


async def publish_dataset(
    *,
    workflow_id: int,
    workflow_name: str,
    readme_md: str,
    tab_results: List[TabResult],
) -> Dict[str, Any]:
    """Write Frictionless datasets to S3: Parquet, CSV and JSON.

    The workflow must be locked: this function cannot be called twice on the
    same workflow at the same time.

    Return a Frictionless Data spec
    """
    try:
        last_revision = _get_latest_revision(workflow_id)
        revision = last_revision + 1
    except FileNotFoundError:
        last_revision = None
        revision = 1

    s3_prefix = f"wf-{workflow_id}/r{revision}"
    url_prefix = f"{settings.API_URL}/v1/datasets/{workflow_id}-{slugify(workflow_name)}/r{revision}"

    s3.remove_recursive(s3.DatasetsBucket, s3_prefix + "/")

    if readme_md:
        # Don't publish empty README.md: boto3 can't upload empty files to minio
        # https://github.com/minio/minio/issues/11245
        s3.put_bytes(
            s3.DatasetsBucket, s3_prefix + "/README.md", readme_md.encode("utf-8")
        )

    resources = []
    for tab_result in tab_results:
        resources.extend(
            _publish_tab_resources(
                url_prefix=url_prefix, s3_prefix=s3_prefix, tab_result=tab_result
            )
        )

    frictionless_datapackage_spec = _publish_datapackage(
        url_prefix=url_prefix,
        s3_prefix=s3_prefix,
        workflow_id=workflow_id,
        workflow_name=workflow_name,
        revision=revision,
        resources=resources,
    )

    # Mark this revision the "current" one.
    s3.copy(
        s3.DatasetsBucket,
        f"wf-{workflow_id}/datapackage.json",
        f"{s3.DatasetsBucket}/{s3_prefix}/datapackage.json",
    )

    # Mark previous revision for deletion
    if last_revision:
        for key in s3.list_file_keys(
            s3.DatasetsBucket, f"wf-{workflow_id}/r{last_revision}/"
        ):
            expires = datetime.datetime.utcnow() + DELAY_FROM_DATASET_EXPIRED_TO_DELETED
            # [adamhooper, 2021-07-22] tested on S3: overwrite updates LastModified
            # [adamhooper, 2021-07-22] tested on GCS: overwrite updates LastModified
            # tested on minio: nope. https://github.com/minio/minio/issues/12777
            # So let's just overwrite some metadata, as a placeholder until
            # https://www.pivotaltracker.com/story/show/178977059
            s3.copy(
                s3.DatasetsBucket,
                key,
                f"{s3.DatasetsBucket}/{key}",
                MetadataDirective="REPLACE",
                Metadata={"cjw-expires": expires.isoformat() + "Z"},
            )

    return frictionless_datapackage_spec
