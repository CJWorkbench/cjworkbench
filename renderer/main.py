from cjwstate import rabbitmq
from cjworkbench.pg_render_locker import PgRenderLocker
from .render import handle_render


async def main_loop():
    """
    Run fetchers and renderers, forever.
    """
    async with PgRenderLocker() as pg_render_locker:

        @rabbitmq.manual_acking_callback
        async def render_callback(message, ack):
            return await handle_render(message, ack, pg_render_locker)

        connection = rabbitmq.get_connection()
        connection.declare_queue_consume(rabbitmq.Render, render_callback)
        # Run forever
        await connection._closed_event.wait()
