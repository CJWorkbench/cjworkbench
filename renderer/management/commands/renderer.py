import asyncio
import logging
import os
from django.core.management.base import BaseCommand
from ...main import main_loop


logger = logging.getLogger(__name__)


def exit_on_exception(loop, context):
    logger.error('Exiting because of unhandled error: %s\nContext: %r',
                 context['message'], context,
                 exc_info=context.get('exception'))
    os._exit(1)


async def main():
    loop = asyncio.get_running_loop()
    loop.set_exception_handler(exit_on_exception)
    await main_loop()


class Command(BaseCommand):
    help = 'Continually render stale workflows'

    def handle(self, *args, **options):
        asyncio.run(main())
