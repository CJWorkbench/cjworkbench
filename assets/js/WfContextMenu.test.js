import React from 'react'
import WfContextMenu  from './WfContextMenu'
import renderer from 'react-test-renderer'

it('WfContextMenu renders correctly', () => {
  const tree = renderer.create(
    <WfContextMenu removeModule={ () => {} } />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});





