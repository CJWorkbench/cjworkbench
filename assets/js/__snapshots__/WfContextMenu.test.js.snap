// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WfContextMenu renders correctly 1`] = `
<WfContextMenu
  removeModule={[Function]}
>
  <UncontrolledDropdown>
    <Dropdown
      isOpen={false}
      tag="div"
      toggle={[Function]}
    >
      <div
        className="dropdown"
      >
        <DropdownToggle
          aria-haspopup={true}
          className="context-button"
          color="secondary"
          data-toggle="dropdown"
        >
          <Button
            aria-expanded={false}
            aria-haspopup="true"
            className="context-button"
            color="secondary"
            data-toggle="dropdown"
            onClick={[Function]}
            tag="button"
          >
            <button
              aria-expanded={false}
              aria-haspopup="true"
              className="context-button btn btn-secondary"
              data-toggle="dropdown"
              onClick={[Function]}
              type="button"
            >
              <div
                className="button-icon icon-more"
              />
            </button>
          </Button>
        </DropdownToggle>
        <DropdownMenu
          tag="div"
        >
          <div
            aria-hidden={true}
            className="dropdown-menu"
            role="menu"
            tabIndex="-1"
          >
            <DropdownItem
              className="dropdown-menu-item"
              onClick={[Function]}
              tag="button"
              toggle={true}
            >
              <button
                className="dropdown-menu-item dropdown-item"
                onClick={[Function]}
                tabIndex="0"
                type="button"
              >
                <i
                  className="icon-share "
                />
                <span
                  className="t-d-gray content-3 ml-3"
                >
                  Share
                </span>
              </button>
            </DropdownItem>
            <DropdownItem
              className="dropdown-menu-item"
              onClick={[Function]}
              tag="button"
              toggle={true}
            >
              <button
                className="dropdown-menu-item dropdown-item"
                onClick={[Function]}
                tabIndex="0"
                type="button"
              >
                <i
                  className="icon-bin"
                />
                <span
                  className="t-d-gray content-3 ml-3"
                >
                  Delete
                </span>
              </button>
            </DropdownItem>
          </div>
        </DropdownMenu>
      </div>
    </Dropdown>
  </UncontrolledDropdown>
</WfContextMenu>
`;
